src/storage/engine/mod.rs

```diff
@@ -101,6 +103,21 @@ pub trait Engine: Send + Debug {
         }
     }
 
+    fn future_snapshot(
+        &self,
+        ctx: &Context,
+    ) -> Box<Future<Item = Box<Snapshot + 'static>, Error = Error> + Send> {
```

Comment:
can we use impl Future here?

Reply:
nop, for two reasons:

1. it is in a trait, and impl Trait in trait is not supported for now.
2. it have different return types (although both of them implements `Future<>`) and this is not supported as well currently.